selfdrive/updated.py

```
@@ -1,57 +1,267 @@
+libc = ffi.dlopen(None)
 
-    # download application update
-    try:
-      r = subprocess.check_output(NICE_LOW_PRIORITY + ["git", "fetch"], stderr=subprocess.STDOUT).decode('utf8')
-    except subprocess.CalledProcessError as e:
-      cloudlog.event("git fetch failed",
-        cmd=e.cmd,
-        output=e.output,
-        returncode=e.returncode)
-      time.sleep(60)
-      continue
-    cloudlog.info("git fetch success: %s", r)
+# **** helper functions ****
 
-    # Write update available param
-    try:
-      cur_hash = subprocess.check_output(["git", "rev-parse", "HEAD"]).rstrip()
-      upstream_hash = subprocess.check_output(["git", "rev-parse", "@{u}"]).rstrip()
-      params.put("UpdateAvailable", str(int(cur_hash != upstream_hash)))
-    except:
-      params.put("UpdateAvailable", "0")
+def wait_between_updates():
+  if SHORT:
+    time.sleep(10)
+  else:
+    time.sleep(60 * 10)
+
+def link(src, dest):
+  # Workaround for the EON/termux build of Python having os.link removed.
+  return libc.link(src.encode(), dest.encode())
+
+def run(cmd, cwd=None):
+  return subprocess.check_output(cmd, cwd=cwd, stderr=subprocess.STDOUT, encoding='utf8')
+
+# **** meaty functions ****
+
+def init_ovfs():
+  if os.path.ismount(OVERLAY_MERGED):
+    # We had to have the overlay lock to get this far, so this must be a stale
+    # mount from a terminated updated process.
+    run(["umount", OVERLAY_MERGED])
+
+  cloudlog.info("preparing new staging area")
+  if os.path.isfile(os.path.join(BASEDIR, ".update_succeeded")):
+    os.remove(os.path.join(BASEDIR, ".update_succeeded"))
+  if os.path.isdir(STAGING_ROOT):
+    shutil.rmtree(STAGING_ROOT)
+
+  for dirname in [STAGING_ROOT, OVERLAY_UPPER, OVERLAY_METADATA, OVERLAY_MERGED, FINALIZED]:
+    os.mkdir(dirname, 0o755)
+  if not os.lstat(BASEDIR).st_dev == os.lstat(STAGING_ROOT).st_dev:
+    cloudlog.error("base and staging directories are on different filesystems, not valid for overlay FS!")
+    exit(1)
```

Maybe throw a RunTimeError instead of cloudlog + exit. If I look at other uses of cloudlog.error they are all recoverable errors.

---

Can do.